Restore timer state on restart

e2m32 opened the pull request:
## Breaking Change:

Timers will resume on HA start. Timer will no longer return to `IDLE` state unless the time for the timer to have `finished` has passed while HA was down.

## Description:

Default behavior is to restore the timer on restart with no grace period. Meaning that the timer will only continue if there is remaining time left on the timer when HA starts.

If timer `finished` event should have fired during the time that HA was down, the `restore_grace_period` is checked. By default there is no grace period. If the time from when the `finished` event should have been fired and the current time is greater than the `restore_grace_period`, then no event is fired and the timer is reset to it's `IDLE` state.

Optional: `restore: false` in config allows the timer state to be set to `idle` after a restart (this restores the current behavior for a timer).
Optional: `restore_grace_period: "00:15:00"' This allows for additional time (15 minutes in example) to be allowed on the timer if the timer should have fired the `finished` event while HA was down.

For a more detailed description with examples look here: https://github.com/e2m32/hass_custom_timer

**Related issue (if applicable):** fixes #
https://github.com/home-assistant/home-assistant/issues/26713


**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#11117

## Example entry for `configuration.yaml` (if applicable):
```yaml
timer:
  laundry:
    duration: '00:01:00'
    restore: true
    restore_grace_period: '00:15:00'
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

e2m32 committed:
Updated timer component to be able to be restored on HA restart

- Previous timer would be reset to `IDLE` state regardless of the state
  when HA stopped. This update defaults to that behavior, but has
  an optional `restore` flag that will restore the timer to the previous
  state. If the timer was `active` it will resume the timer (including
  the time that HA was down). If the timer should have `finished`
  while HA was down, it will immediately fire the `finished` event
  as long and the current time is within the `restore_grace_period`
  from when the event should have fired. If the current time is not
  within the `restore_grace_period` the event will be ignored and
  the timer will be reset to the `IDLE` state.

e2m32 committed:
Lint Update

homeassistant commented:
Hi @e2m32, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#28593).

Once you do that we will be able to review and accept this pull request.

Thanks!

pvizeli commented:
I don't think that need an new config option? They can just restore if it is possible?

e2m32 commented:
I agree @pvizeli, being my first pull request I was a bit sheepish about making a breaking change and in a discussion on the community forum there were some comments about not wanting all timers to be restored on a restart. So I figured I would just keep the default behavior the same with the option to restore.

If we make the `restore:true` the default behavior, what are your thoughts on the default `restore_grace_period`?

mmakaay commented:
Thank you for this pull request. It fixes the exact issue I am running into right now.
As for default behavior: I would vote for always store and unlimited grace period (so timers will always trigger their intended jobs). If a value is required, then maybe a day is the cleanest, since that might generally best match a day by day pattern in which the timers are triggered. E.g. If I use this for my shower light timer, then chances are really high that I would start a new timer within the next 24 hours.

tdejneka commented:
>  I would vote for always store and **unlimited grace period** (so timers will always trigger their intended jobs)
I would _not_ vote for a default **unlimited** grace period. Not all timer applications must trigger after an extended outage. A timer can become _stale_. In other words, by the time Home Assistant restarts (possibly after an extended power failure) the timer's purpose has expired. The automation it executes might perform an operation that is no longer needed (i.e. open a window shutter in 15 minutes may no longer make sense in 1 hour and 15 minutes).

It's vital that the user decide the duration of a timer's grace period (with a default value of none or very little) and not to simply default to an unlimited duration.

mmakaay commented:
Choosing a default grace period is quite hard. I can come up with many examples that would plead for different default values. For me, any default grace period would work though. I am highly in favor of the "explicit before implicit" rule, and will define grace periods explicitly in my config anyway.

As for your proposal:
Wouldn't a grace period of "none" (i.e. zero?) be a bit strange?
Wouldn't that effectively counter-act the default `restore: true` value?
A short time makes more sense. For example 5 minutes, likely covering regular server restarts and device reboots.

tdejneka commented:
> Wouldn't a grace period of "none" (i.e. zero?) be a bit strange?

The concept of a grace period for a timer is strange. Effectively, it allows a timer to run for _more time than its intended duration_.

Let's say we have a 15 minute timer with the proposed default 15 minute grace period.
- Timer starts at 10:00. It should finish at 10:15.
- Power failure at 10:10. Home Assistant stops.
- Power restored and Home Assistant restarts at 10:29.
- 10:29 is less than 10:30 (10:15 + 15 minutes grace period) so the timer will be allowed to trigger its `finished` event.
- What was supposed to be a **15** minute timer, is effectively allowed to run for **29** minutes.


In contrast to the `time` platform, where triggering occurs at a specific point in time, timers are often used to provide 'temporal spacing' (delay) between events. For example, event B should occur 10 minutes after event A. Allowing for a default grace period of 15 minutes creates the possibility of event B occurring not 10 but almost 25 minutes later. Upon restart, the automation associated with this resumed timer will be triggered well past the time it should have occurred with potentially undesirable (or worse) consequences.

This PR is greatly appreciated for resuming timers that _still have a remaining duration to complete_. If a 1 hour timer is interrupted for 45 minutes and, upon restart is determined to have 15 minutes left, there's nothing unusual or controversial about allowing it to finish. **It's total run-time remains its specified duration.**

On the other hand, timers that _expire during the outage_ but are then _resumed_, because of a default 15 minute grace period, that's new and unusual behavior. It may come as a surprise to users who simply added `restore: true` to the timer's configuration. That's why I suggested the default should be `0` to allow timers to expire naturally (during the outage). If someone wishes to have them reincarnated at startup, then they can explicitly change the timer's grace period from `0` to whatever they desire.

mmakaay commented:
Yes, you're absolutely right. The term "grace period" took a wrong turn in my head. Your example cleared that up. A grace period of zero does indeed make sense as a default and it has my vote.

Note however: there might be some cases where users think they are experience buggy behavior. Think for example about this basic pattern:

* A light is turned on by a motion sensor
* A timer of 5 minutes starts
* When the timer triggers, the light is turned off

During most of the timer period, a restart can been done without any issues. The timer  can continue after the restart at the point where it would have been without the restart and eventually it will trigger.
However, a restart at 4:55 becomes interesting. If the restart takes more than 5 seconds, using a grace period of zero, the timer would not trigger anymore. Result: the light stays on, possibly unexpected for the user who expected and wanted the light to go off.

I don't think that there is a default grace period that will fit every use case.
I do think that clear documentation and examples can take away possible confusion.

e2m32 commented:
@tdejneka you make a good point. We have no idea what the user is going to be using the timer for, so we should assume the expected behavior of not surpassing the set time. The user might wonder why the timer `finished` event didn't fire when they expected it to and when they dig into it, they will realize that HA was down during that time and couldn't fire the event which will probably be what anyone would expect. I like having the option of the grace period and it is set per timer, so the user can decide which timers should have it. 

In my personal use case, I am often using a timer to turn something off, so I want the grace period to make sure that that device gets turned off... but if HA is down for a long time, I will have manually intervened by that point. But that may not be everyone's use case. That's the beauty of HA, you can use these components to do anything. 😄 

I'll update with `restore`: true and `restore_grace_period` : 0 as the defaults

mmakaay commented:
Love it :-)
One of my bit and important use cases is keeping the wife happy. While she is home, walking around and triggering lights all over the place, she doesn't like it when lights stay turned on, because a certain somebody was away from home and tinkering with HA (since it's addicting) ;-)

e2m32 committed:
Changed defaults for restore

- Got good feedback from @pvizeli,  @mmakaay, and @tdejneka
  on setting `restore`: true and `restore_grace_period`: 0 as the
  defaults.
- Fixed default `duration` not being set
- Some code cleanup

pnbruckner commented:
I'm strongly against changing the current behavior in the default case.

People have been creating and fine tuning many automations using timers that all assume a timer will start as idle, regardless of what it was doing before HA stopped. Some of these can be quite complex where the timer is manipulated in multiple ways, much more than simply starting them and waiting for them to finish. What if some event happens while HA is down that would have canceled the timer? A grace period can't know that.

I'm not against adding useful functionality, but it _must_ be done in a way that doesn't break the thousands upon thousands of working automations.

Swamp-Ig commented:
Agree with Phil.

Maybe a flag to persist the state?

  persist: True

On Thu, 21 Nov 2019 at 11:19, Phil Bruckner <notifications@github.com>
wrote:

> I'm strongly against changing the current behavior in the default case.
>
> People have been creating and fine tuning many automations using timers
> that all assume a timer will start as idle, regardless of what it was doing
> before HA stopped. Some of these can be quite complex where the timer is
> manipulated in multiple ways, much more than simply starting them and
> waiting for them to finish. What if some event happens while HA is down
> that would have canceled the timer? A grace period can't know that.
>
> I'm not against adding useful functionality, but it *must* be done in a
> way that doesn't break the thousands upon thousands of working automations.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/28593?email_source=notifications&email_token=ABQGMTNA6QSQIJFUXQHUYE3QUX42RA5CNFSM4JJ3ORMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEY2L3Q#issuecomment-556901870>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABQGMTPYB277GLMVSG5DVG3QUX42RANCNFSM4JJ3ORMA>
> .
>

e2m32 committed:
Merge with 0.105.0.dev0, Allow for >24hr timers

- Timers > 24 hours now display in lovelace and
  are able to be restored.

e2m32 committed:
Merge remote-tracking branch 'upstream/dev' into timer-restore-on-restart

# Conflicts:
#	homeassistant/components/timer/__init__.py

balloob commented on homeassistant/components/timer/__init__.py:
```diff
@@ -69,6 +78,16 @@ def _none_to_empty_dict(value):
         return {}
     return value
 
+def _time_str(time_value):
```
Let's not do this. We have a specific way people specify durations, and that should be the way it is done. We should also try not to put too much into a single PR. So let's remove this.

e2m32 replied:
If we don't do this, timers > 24 hours will not be restored correctly and they will not display in lovelace. 

Try setting a timer > 24 hours in the current release and you will see what I mean. The countdown doesn't display at all. `str(datetime)` for times > 24 hours returns something like this: `D days, HH:MM:SS` If you have a more elegant way to make it `HH:MM:SS` for times greater than 24 hours, we should do that then. Or if we can update the lovelace code to handle the `D days, HH:MM:SS` format, then I will do that (not sure where the code is that handles that though) and update the restore code to handle it as well. I'm open to any suggestions. Thanks @balloob!

databoy2k replied:
Just to throw in my two cents, since I'm the guy who found this bug, it's probably not out of the question that those of us with long timer durations are going to be using this and you should absolutely assume that timers surviving reboots will be >24 hours. Mine are for a week (defined as "168:00:00" in the timer.start duration, but as pointed out being converted to 7 days, 0 hours and causing the lovelace display of the timers to error out) to track chores actually being done.

balloob replied:
Why are we even using `str(datetime)` to begin with? That's not how you serialize a datetime. You nee to use `datetime_value.total_seconds()`.

pnbruckner replied:
Did you mean `datetime_value.timestamp()`?

balloob replied:
Ah, sorry I was thinking of a duration. I should have written `timedelta_value.total_seconds()`. For datetime we should serialize it as `datetime_value.isoformat()`

e2m32 replied:
I get using the isoformat, but the End value is included in the timer attributes so that the timer can be restored. Users will see this when they look at the state card. It might be a little confusing/less pretty with a UTC time.

Currently displayed:
`2020-01-25 09:47:39.234619-05:00`
Isoformat:
`2020-01-25T14:47:39.234619`
I thought that seeing your time with your timezone information would be more clear to the user. At first glace you might think that the timer is wrong.

Similarly, if we store the duration as seconds then the Duration attribute will display the number of seconds in the state card in lovelace instead of the pretty `00:00:00` that matches how users are likely to enter this information in the config. Also, I just tested it and the lovelace countdown can't interpret longer than 24 hours, so no countdown is displayed when the timer is active. My `_time_str` worked.

e2m32 replied:
I'm going to leave it the way it was and make a new PR for this to allow for timers > 24 hours. We can table the discussion for that PR.

balloob commented on homeassistant/components/timer/__init__.py:
```diff
@@ -29,14 +29,23 @@
 
 DEFAULT_DURATION = 0
+DEFAULT_RESTORE = False
+DEFAULT_RESTORE_GRACE_PERIOD = timedelta(minutes=15)
```
We decided in the comments to not have a default grace period.

balloob commented on homeassistant/components/timer/__init__.py:
```diff
@@ -29,14 +29,23 @@
 ENTITY_ID_FORMAT = DOMAIN + ".{}"
 
 DEFAULT_DURATION = 0
+DEFAULT_RESTORE = False
```
This is good. This also makes it not a breaking change. So please remove that part of the description.

e2m32 replied:
Yeah, there was a bit of a back and forth on this PR whether it should be a breaking change or not. I wanted to wait to update the description until a decision was made.

balloob commented on homeassistant/components/timer/__init__.py:
```diff
@@ -1,5 +1,5 @@
 """Support for Timers."""
```
Please do not change the file mode. Revert back to 644.

e2m32 replied:
You are referring to the chmod file permissions on this right? Windows probably overwrote them at some point. I will correct.

balloob replied:
yeah

balloob commented on homeassistant/components/timer/__init__.py:
```diff
@@ -161,6 +186,8 @@ class TimerStorageCollection(collection.StorageCollection):
         data = self.CREATE_SCHEMA(data)
         # make duration JSON serializeable
         data[CONF_DURATION] = str(data[CONF_DURATION])
+        data[CONF_RESTORE] = str(data[CONF_RESTORE])
```
Don't store booleans as strings…

balloob commented on homeassistant/components/timer/__init__.py:
```diff
@@ -185,6 +214,15 @@ def __init__(self, config: typing.Dict):
         self.editable = True
         self._state = STATUS_IDLE
         self._remaining = config[CONF_DURATION]
+        self._restore = config[CONF_RESTORE] if config[CONF_RESTORE] is not None \
```
This doesn't work. `config[CONF_RESTORE]` raises a `KeyError` because the data is not in the config.

balloob replied:
You should replace this line with `config.get(CONF_RESTORE, DEFAULT_RESTORE)`

balloob commented on homeassistant/components/timer/__init__.py:
```diff
@@ -185,6 +214,15 @@ def __init__(self, config: typing.Dict):
+                        else DEFAULT_RESTORE
+        if self._restore:
+            self._restore_grace_period = config[CONF_RESTORE_GRACE_PERIOD] \
+                                    if config[CONF_RESTORE_GRACE_PERIOD] is not None \
```
Same. Also please do not use multi-line inline if-statements.

balloob commented on homeassistant/components/timer/__init__.py:
```diff
@@ -220,9 +258,14 @@ def state(self):
+            ATTR_REMAINING: _time_str(self._remaining),
+            ATTR_RESTORE: str(self._restore),
+            ATTR_RESTORE_GRACE_PERIOD: str(self._restore_grace_period),
+            ATTR_END: str(self._end.replace(tzinfo=timezone.utc).astimezone(tz=None)) \
```
Please limit the PR only to the changes that are in the description. Remove this.

e2m32 replied:
You're referring to the timer > 24 hours issue right? Using `_time_str()` here instead of `str()`, use a separate pull request and come up with a better solution? So I should revert this back to `str()` for now and issue a new PR for allowing for longer timers?

balloob commented on homeassistant/components/timer/__init__.py:
```diff
@@ -232,12 +275,60 @@ def unique_id(self) -> typing.Optional[str]:
+        # Check for previous recorded state
         state = await self.async_get_last_state()
-        self._state = state and state.state == state
+        if state is not None:
```
Please use a guard clause. `if state is None: self._state = STATUS_IDLE; return`

balloob commented on homeassistant/components/timer/__init__.py:
```diff
@@ -232,12 +275,60 @@ def unique_id(self) -> typing.Optional[str]:
-        self._state = state and state.state == state
+        if state is not None:
+            for check_status in VIABLE_STATUSES:
+                if state.state == check_status:
```
Use a guard clause. `if state.state != check_status: continue`

balloob commented on homeassistant/components/timer/__init__.py:
```diff
@@ -232,12 +275,60 @@ def unique_id(self) -> typing.Optional[str]:
+                                    if not state.attributes.get(ATTR_END) == "None" \
+                                       and state.attributes.get(ATTR_END) is not None \
+                                    else None
+                    except:
```
Please specify the exact error that you're catching .

balloob commented on homeassistant/components/timer/__init__.py:
```diff
@@ -232,12 +275,60 @@ def unique_id(self) -> typing.Optional[str]:
+                        self._remaining = timedelta()
+                    try:
+                        self._end = datetime.strptime(state.attributes.get(ATTR_END), "%Y-%m-%d %H:%M:%S%z") \
+                                    if not state.attributes.get(ATTR_END) == "None" \
```
We should never ever end up with a string `"None"`.

balloob commented on homeassistant/components/timer/__init__.py:
```diff
@@ -232,12 +275,60 @@ def unique_id(self) -> typing.Optional[str]:
+                            self._remaining = timedelta(hours=remaining_dt[0],
+                                                        minutes=remaining_dt[1],
+                                                        seconds=remaining_dt[2])
+                        except:
```
No blank except, always specify exact exceptions.

balloob reviewed (changes requested):
Please add tests. Make sure you include tests that use the current data format and process it in the new approach and add tests that provide data in the new format and make that work. Test all different permutations of the options that you've added.

e2m32 committed:
Merge remote-tracking branch 'upstream/dev' into timer-restore-on-restart

e2m32 committed:
Updated Default and FIELDS, removed _time_str

- Corrected permissions
- Updated Restore Default to 0
- Updated CREATE_FIELDS
- Updated UPDATE_FIELDS
- Removed _time_str()
  - This allows for timers > 24 hours.
  - Will play around with a better way for a new PR.

e2m32 committed:
Merge remote-tracking branch 'upstream/dev' into timer-restore-on-restart

e2m32 committed:
Merge remote-tracking branch 'upstream/dev' into timer-restore-on-restart
